sanitize looping over equipment arrays

neolit123 opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [x] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->

3 commits that sanitize loops that iterate global equipment arrays (yikes).
because reading adjacent memory can crash from time to time.

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->

- use MAX_TANK_INFO and MAX_WS_INFO (new) everywhere

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

Fixes #1401

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

na

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->

`Equipment: fix bug with the maximum number of allowed tanks [#1401]`

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

na

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->

@dirkhh @atdotde @bstoeger @p-neves

bstoeger commented:
I'm not very fond about the `MAX_*` + searching for first `NULL` pattern. Why not a pointer-to-array + length pair. The array could then be `realloc()`ed on need. Of course one could also simply use a `QVector`, which can be accessed from C via helper (`add_x()`, `get_x()`, `delete_x()`) functions.

In `core/equipment.c` we find the comment
```
/* placeholders for a few functions that we need to redesign for the Qt UI */
```
I read this as a plan to move this functionality to Qt?

bstoeger commented on qt-models/weightsysteminfomodel.cpp:
```diff
@@ -80,7 +80,7 @@ WSInfoModel::WSInfoModel() : rows(-1)
 	setHeaderDataStrings(QStringList() << tr("Description") << tr("kg"));
 	struct ws_info_t *info = ws_info;
-	for (info = ws_info; info->name; info++, rows++) {
+	for (info = ws_info; info->name && info < ws_info + MAX_WS_INFO; info++, rows++) {
```
`info` is initialized twice.

bstoeger commented on qt-models/weightsysteminfomodel.cpp:
```diff
@@ -118,8 +118,7 @@ void WSInfoModel::update()
 	struct ws_info_t *info = ws_info;
-	for (info = ws_info; info->name; info++, rows++)
-		;
+	for (info = ws_info; info->name && info < ws_info + MAX_WS_INFO; info++, rows++);
```
`info` is initialized twice.

bstoeger commented on qt-models/weightsysteminfomodel.cpp:
```diff
@@ -98,7 +98,7 @@ void WSInfoModel::updateInfo()
 	beginRemoveRows(QModelIndex(), 0, this->rows);
 	endRemoveRows();
 	rows = -1;
```
I don't like `rows == -1` meaning no rows.

neolit123 replied:
i left this outside of the scope of this PR.

bstoeger commented on qt-models/weightsysteminfomodel.cpp:
```diff
@@ -98,7 +98,7 @@ void WSInfoModel::updateInfo()
 	beginRemoveRows(QModelIndex(), 0, this->rows);
```
Here and elsewhere: To reset the model, there is `beginResetModel()`/`endResetModel()`.

neolit123 replied:
yes, this can work.

neolit123 commented:
update:
- i've addressed the two double-initialized variables.
- added CHANGELOG entry for the bugfix.

> I'm not very fond about the MAX_* + searching for first NULL pattern. Why not a pointer-to-array + length pair. The array could then be realloc()ed on need. Of course one could also simply use a QVector, which can be accessed from C via helper (add_x(), get_x(), delete_x()) functions.

it can be improved for sure, but like @dirkhh implies it's probably not worth the effort ATM.
we have 70+ issues in GitHub, so we should definitely focus more on bug fixes instead of the refactoring i would say.

> In core/equipment.c we find the comment
/* placeholders for a few functions that we need to redesign for the Qt UI */
I read this as a plan to move this functionality to Qt?

so as a little bit of a backstory here - we wanted the core of Subsurface to be a C library at one point. like people wrapping it around the GUI framework they want. of course, this kind of became difficult the moment we started introducing C++/Qt helpers.

so nowadays we don't have the Subsurface C library, but we still want to keep the core separate from C++ as much as possible, i guess.

dirkhh commented:
This looks sane and correct. BTW, the 'Fixes #1401' needs to be in one of the commits, not in the PR
